### PR TITLE
Bump htslib, bcftools, and samtools to 1.12

### PR DESCRIPTION
**Pull request by jmarshall**
Bump the three formulae, and also make minor improvements to _htslib.rb_ and _bcftools.rb_ as described in the commit messages.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

~~For bcftools, `audit --strict` complains that “GPL-3.0” should specify “-only” or “-or-later”. The bcftools source code is itself MIT/Expat-licensed, but when the executable is linked with the GNU Scientific Library, as it is here, the resulting executable is GPL-licensed. I have copied the license tag from _Formula/gsl.rb_, so determining whether both license tags should be only or or-later should really be done by someone knowledgable about GSL, e.g., brew's _Formula/gsl.rb_ maintainer.~~

### CODE REVIEW

File: Formula/bcftools.rb
Comment (carlocab) [inline Formula/bcftools.rb:8]: Does `bcftools` link statically with `gsl`? (If it does, it probably shouldn't.) I believe that's the only time linking affects the license. [1] Otherwise the listed license can stay MIT.

[1] There is some contention over this issue.

File: Formula/bcftools.rb
Comment (carlocab) [inline Formula/bcftools.rb:23]: Does `brew linkage bcftools` show no linkage with `xz`?

**Review by carlocab — COMMENTED** [review]
Thanks for your PR, @jmarshall. I have a few questions.

Also:

>brew's Formula/gsl.rb maintainer.

There are no maintainers assigned to specific formulae, so it would be great if you could help determine whether this is `-or-later` or `-only`. Looking in the source headers for the phrases `at your option` or `later {V,v}ersion` might help.

File: Formula/bcftools.rb
Comment (jmarshall) [reply Formula/bcftools.rb:23]: Indeed. The only homebrew libraries it lists are
```
Homebrew libraries:
  /usr/local/opt/gsl/lib/libgsl.25.dylib (gsl)
  /usr/local/opt/htslib/lib/libhts.3.dylib (htslib)
```

File: Formula/bcftools.rb
Comment (jmarshall) [reply Formula/bcftools.rb:8]: It links dynamically with `gsl`. The bcftools [LICENSE](https://github.com/samtools/bcftools/blob/1.12/LICENSE) and [INSTALL](https://github.com/samtools/bcftools/blob/1.12/INSTALL#L132-L159) documentation make it clear that the bcftools authors consider that this makes the dynamically-linked `bcftools` executable GPL-licensed.

If it is the belief of the Homebrew maintainers that the listed license should stay MIT, then they should revert this part of the change.

Comment (carlocab) [reply Formula/bcftools.rb:8]: I'm happy to defer to the `bcftools` documentation on this. If it's their opinion their software is GPL-licensed when linking with `gsl` (even dynamically), then we should respect that.

However, I would like to resolve whether this is `-or-later` or `-only`.

Comment (jmarshall) [reply Formula/bcftools.rb:8]: I may have time available to investigate this in due course. Alternatively you could resolve this question by checking the GSL source yourself; or the _gsl.rb_ license line was added by @whoiswillma in 8803565ddce998371c13f1f76bd2f9f2cf12bda6 so perhaps he is the local knowledgeable person.

Comment (whoiswillma) [reply Formula/bcftools.rb:8]: I added license information using an automated script last summer in order to bootstrap the license DSL. My apologies but I don't know much about `gsl` specifically

**Comment by carlocab** [comment]
You also modify `bcftools` in 4b2dcf74d8068091db8484a2029df528aa586b9f. Can you move that change to the commit in 2fe09f7f8b43042986d630a7bef3a10d1b9259be, so we can get rid of the automerge skip flag?

There should only be one formula-file modified per commit (and one commit per formula-file modified). (Ref. [docs](https://docs.brew.sh/Formula-Cookbook#push))

Thanks!

**Comment by jmarshall** [comment]
Done. (You know, you could have adjusted the commits to your taste in less time than it took you to explain to an external volunteer contributor how you wanted them to be arranged.)

**Comment by carlocab** [comment]
Sure, but the quick way to do that is a `git reset`. This destroys commit authorship data, and I wanted to avoid that. I could've also done a sequence of amending commits and rebasing to preserve the authorship of your commits, but then that takes about as much time, if not more, than the explanation.

Plus, actually writing down the explanation might've saved me the time of doing the above again in the future.

**Review by carlocab — APPROVED** [review]
Thanks, @jmarshall. Merging.

**Pull request closed without merging**
